code/ui/manager/src/components/preview/preview.stories.tsx

```diff
@@ -160,3 +160,40 @@ export const WithTabsHidden = () => (
     }}
   </Consumer>
 );
+
+export const WithToolbar = () => <Preview {...previewProps} />;
+
+export const WithToolbarHidden = () => (
+  <Consumer>
+    {({ api }: Combo) => {
+      return (
+        <Preview
+          {...previewProps}
+          options={{ ...previewProps.options, showToolbar: false }}
+          api={{ ...api, getElements: () => ({}) }}
+        />
+      );
+    }}
+  </Consumer>
+);
+
+export const WithToolbarHiddenFromConfig = () => {
+  useEffect(() => {
+    addons.setConfig({
+      toolbar: {
+        showToolbar() {
+          return false;
+        },
+      },
+    });
+
+    return () => {
+      addons.setConfig({
+        toolbar: {
+          showToolbar: null,
+        },
+      });
+    };
```

Comment:
Wouldn't this affect the true manager rather than the mocked one provided by the `ManagerProvider` above?

Reply:
Absolutely. This is unavoidable for now as the addons store is a global singleton. As the store `setConfig` function only adds to the current store, I need to actively remove the function for the story to have _reasonable_ side effects (but side effects nonetheless). If I had had a default state for that store, I'd have had much worse side effects.

I see two ways of resolving this:
* Providing addons through the manager React context instead of as a singleton (massively breaking change)
* Editing the addon store to add a save/restore mechanism for tests, either through
  * Explicit functions, one to snapshot/save the state, and one to restore the state to the last saved snapshots, which makes it usable by end users with all the  implications
  * A wrapper function for arbitrary code, that saves upon entering and restores upon leaving, which would be more specifically for test stories to use as a decorator

Of course I might have missed that there's already a method to solve my issue in the codebase.

Reply:
Ok, wow, I didn't realise that. That's kind of a mess.

> Providing addons through the manager React context instead of as a singleton (massively breaking change)

It's pretty clear we should do this. I don't know if it's a breaking change to start reading addon info from a provider in a *new* spot we are accessing it (i.e. in this PR). But understand if it feels out of scope of this PR.

@ndelangen any thoughts on this?

--

I have another suggestion, if the above is too much -- which is to make a test-only prop to the `PreviewComponent` for the result of `addons.getConfig()`. 

So the code might look something like:

```ts
const useShowToolbar = (showToolbar: boolean, overrideAddonsConfig: Record<string, any>) => {
  const config = overrideAddonsConfig || addons.getConfig()
}
```

--

In lieu of the either above, I'm not a big fan of a story that messes with the storybook that renders it, that seems like asking for trouble (even when you make sensible efforts to undo those changes). I'd probably advocate something that gives us less coverage :shrug:

Reply:
It would be a breaking change in the sense that users expect it to be global. They expect to be able to call `addons.setConfig` and have the changes reflected.

If my code used the addons store through a manager context instead, with the underlying assumption that the store content is unique to the manager and separated from the global singleton, then using the current global API could sometimes fail to result in a change to the store that the toolbar reads from.

----

That's why I'd prefer to have a pattern built into the store. It lets us build "social unit tests" safely so long as Storybook isn't rendering two stories at the same time (safe assumption to make?).

Snapshotting the store state and restoring it is easy to build within the store. There would be side effects if/when you parallelise story tests, though, so then the pattern could not be used as is. Stories like mine could be flagged to not be parallelised to avoid this issue.

----

Alternatively, detaching the store state could work and could handle the problems of concurrent rendering or concurrent testing:
* Instead of snapshotting the story state and restoring it, the addons store could create a detached, uniquely identified store copy when asked
* In stories that use the addon store, we'd create detached copies identified by the story id
* `addons.setConfig` would need an extra optional param; when passed, it uses the detached copy with the matching id, **or reverts to the main store**
* Internally within SB, where we want support for story tests with addons, we pass the story id to our `addons.setConfig` calls, which results in:
   * In a test story context, the detached store is used and the main store is not impacted
   * In a production environment, the main store is always used

```ts
// preview.stories.tsx
export const WithToolbarHiddenFromConfig = () => {
  useEffect(() => {
    addons.detach('story-id-goes-here-somewhat')
    addons.setConfig({
      toolbar: {
        showToolbar() {
          return false;
        },
      },
    }, 'story-id-goes-here-somewhat');

    return () => {
      addons.cleanup('story-id-goes-here-somewhat');
    };
  }, []);
  return <Preview {...previewProps} />;
};
```

Ultimately this is a variation of what you're proposing, though done at the store level:
* I honestly don't see how else I can pass this test prop to the preview component for it to then pass to the toolbar so it feels easier that way for me
* There's the mild advantage that this store change could be used for more than testing a specific component, or even more than writing tests in general

Reply:
@Sidnioulz I think maybe you misunderstood my intention:

> It would be a breaking change in the sense that users expect it to be global. They expect to be able to call addons.setConfig and have the changes reflected.

> If my code used the addons store through a manager context instead, with the underlying assumption that the store content is unique to the manager and separated from the global singleton, then using the current global API could sometimes fail to result in a change to the store that the toolbar reads from.

What I was thinking is simply that either similar to, or even part of [1] [the `ManagerProvider`](https://github.com/storybookjs/storybook/blob/1dc1ed52c1249a26ee3e0ed2d6d5fb5a33cc14f3/code/ui/manager/src/index.tsx#L40-L46), we wrap the whole manager in a context provider that **connects to the existing global singleton**.

Then, in any *new* code (like this PR), we read from the context rather than the global singleton directly.

Then, in stories (of that new code) we can simply wrap the story in a different provider that mocks the addons API.

[1] We could add the addon config to the `Combo` type: https://github.com/storybookjs/storybook/blob/1dc1ed52c1249a26ee3e0ed2d6d5fb5a33cc14f3/code/lib/manager-api/src/index.tsx#L137-L140

Reply:
Ah! I did miss your point indeed :laughing:

As the story renders a manager via a decorator, we could indeed mock it. I initially tried to do that but couldn't find a way to reach the addon singleton that way. If I read `addons` from the context in my toolbar code, it becomes possible indeed. As the manager is within the story, it's well isolated for concurrent tests/rendering.

I'll give it a go ASAP, thanks Tom!

Reply:
Awesome. @ndelangen is away this week but is sort of the expert on the manager architecture so he'll likely want to weigh in next week also.

Reply:
So the issue here is the fact that this API emits an event, that's send across the channel.
https://github.com/storybookjs/storybook/blob/c7028c7b9fdaa77b224c081a8f187906042098d3/code/lib/manager-api/src/lib/addons.ts#L152-L161

So the act of changing this applies to both the manager api instance running in the preview, but also the manager api instance running in the manager.

Calling the underlaying API directly would not have this problem:
https://github.com/storybookjs/storybook/blob/13843a46438058973ede26fe0c2037ae418c6c4d/code/lib/manager-api/src/modules/layout.ts#L288-L290

I think that would solve this issue, but at the cost of not using the "public API".

Reply:
Thanks Norbert!

So, going forward:
* I'll rewrite the preview tests and FakeProvider to inject the addons store directly
* In my toolbar code, I'll retrieve the API via my React Context Consumer
* Once I have it, I'll call setOptions to avoid the side-effect of emitting events on the channel

Is that correct?

Reply:
@Sidnioulz I'm not so sure.
I'm not so sure there's a need for `useShowToolbar.tsx` to exist.

The setConfig API sets state in the `layout` field in the global storybook state:
https://github.com/storybookjs/storybook/blob/d772d382f8a26ab1d31b876239e3a3613c106e3b/code/ui/manager/src/containers/preview.tsx#L25-L42

This initial state is set (but also persisted to/from localStorage):
https://github.com/storybookjs/storybook/blob/13843a46438058973ede26fe0c2037ae418c6c4d/code/lib/manager-api/src/modules/layout.ts#L281-L292

I'm wondering if we can just amend the container component to pull function and pass it to the Preview component. If that's possible we don't need a new context at all.